Control groups docs (#3436)
openbao/openbao: website/content/docs/concepts/control-groups.mdx

---
description: Control-group policy factors for access to secrets
---

# Control groups

## Overview

In some deployments, it may be desirable to require a "second set of
eyes" to approve access to a secret or to generate a
certificate. Control groups implement second-party authorization by
means of policy "factors" applied to a path.  When a control-group
factor applies to a request, one or more members of the approving
control group must assert authorization before the request can be
executed and secret is released.

## The workflow

The general flow for using control groups is as follows:

> 1. A "control_group" stanza is declared in the policy for a path. 

Reviewer: ```suggestion
1. A `control_group` stanza is declared in the policy for a path. 
```

2. When a request for a path triggers the control group factor, the
   response will be [`wrapped`](response-wrapping.mdx).

> 3. The wrapped response will contain a "token" and an "accessor". The

Reviewer: ```suggestion
3. The wrapped response will contain a `token` and an `accessor`. The
```

   requester needs to forward the accessor to a member of the
   approving group and request authorization. **Please note: the approvers are not

>    notified that they have a pending request.**

Reviewer 2: ```suggestion
   approving group and request authorization. **Please note: OpenBao does not send
   any notification to the approvers to tell them that they have a pending request.**
```

4. The approvers can use the accessor to review the request, view the
   authorization state, and add their authorization.
5. Once approved, the wrapping response token can "unwrap" the request
   (which executes the original request).

> ## Control-group response-wrapping

Reviewer 2: ```suggestion
## Control-group response wrapping
```

Response wrapping is triggered by requesting a path having one or more
applicable control group factors in the policy. The
original request is stored for deferred execution upon unwrapping.

> 1. The original HTTP request is serialized and stored

Reviewer 2: Is this literally a `message/http` document being stored? If not, try:
```suggestion
1. OpenBao makes a copy of the original API request, and stores it
```

2. A new single-use token is generated with the TTL supplied by the factor
3. A new response is generated, with the token ID, accessor, TTL, and
   path stored in the new response's wrap information object

> 4. The new response is returned to the caller

Reviewer 2: Not sure if this is correct technically, but try this:
```suggestion
3. OpenBao responds to the original request. Instead of performing the request directly and providing a response for the original request, the response from OpenBao is a _wrap information object_ that includes the token ID, accessor, a time to live (TTL) value for the token, and the original path.
```

Note that policies can control minimum/maximum wrapping TTLs; see the [policies
concepts page](policies.mdx) for
more information.

The wrapped response contains these fields:

- TTL: The TTL of the response-wrapping token itself, specified in the policy factor.
- Token: The actual token value
- Wrapped Accessor: The token used for reviewing and approving the original request.
- Creation Time: The time that the response-wrapping token was created

> - Creation Path: The API path that was called in the original request

Reviewer 2: This could move earlier into the previous ordered list - see https://github.com/openbao/openbao/pull/3436/changes#r3665705520

## Control-group operations

Via the `sys/wrapping` and `sys/control-group` paths, several
operations can be run against control-group wrapping tokens.

> 

Reviewer 2: suggestion: list these in alphabetical order. Conveniently, this puts _authorize_ first (the most useful operation to explain).

- **Unwrap** (`sys/wrapping/unwrap`)**:** Unwrap the token, executing
  the original request and returning a response, once approved. The
  response that is returned will be an original wire-format response;
  it can be used directly with API clients. Note that the original
  request is processed only when unwrapping an approved request.
- **Review** (`sys/control-group/request`)**:** Using the accessor token,
  view information about the original request (entity, path, payload,
  etc).
- **Authorize** (`sys/control-group/authorize`)**:** Using the
  accessor token, approve the request for deferred execution when unwrapped.

## Control-group factors in ACL policies

> Control groups are declared in a control-group stanza of an ACL policy.

Reviewer 2: ```suggestion
Control groups are declared in a control-group stanza of an access control list (ACL) policy.
```

Is it really more of an _access control policy_? It's more complex than a list.

Author: it is more than a list, but ACL seems to be the term of art. Maybe it's a list of policies.

```hcl
path "pki/issue/*" {
  capabilities = ["read", "update"]

  control_group = {
    factor "pki-approvers" {
      controlled_capabilities = ["update"]
      identity {
		# Control groups whose members can authorize access
		group_names = ["pki-approvers", "security-team"]
		approvals = 1

		# Max duration for which authorization is valid
		ttl = "5d"

		# Allow self-authorization when true
		self_authorization = false
	  }
	}
  }
}
```

Control Groups can verify the following factors:

> - `Identity Groups` - Require an authorizer to be in a specific set of

Reviewer 2: Did you mean:
```suggestion
- _Identity groups_ - Require an authorizer to be in a specific set of
```

  identity groups. The `self_authorization` property indicates
  whether the original requester can self-authorize if they belong to
  the control group (default is false).

### Controlled capabilities

Control group factors can be configured to trigger the control group workflow
on specific capabilities. This is done with the `controlled_capabilities` field.
Not specifying the `controlled_capabilities` field will necessitate the factor to be
checked for all operations to the specified policy path. The `controlled_capabilities`
field can differ per factor, so that different factors can be required for different
operations.

Finally, the capabilities in the `controlled_capabilities` stanza must be a subset of the
`capabilities` specified in the policy itself. For example, a policy giving only `read` access to
the path `secret/foo` cannot specify a control group factor with `list` as a controlled capability.

Please see the following section for example ACL Policies.

> ### Sample ACL policies

Reviewer 2: I might write:
```suggestion
### Sample policies
```

```hcl
path "secret/foo" {
    capabilities = ["read"]
    control_group = {
        factor "ops_manager" {
            identity {
                group_names = ["managers"]
                approvals = 1
            }
        }
    }
}
```

The above policy grants `read` access to `secret/foo` only after one member of

> the "managers" group authorizes the request.

Reviewer 2: Is this correct:
```suggestion
The above policy grants `read` access to `secret/foo` only after one member of
the "managers" group authorizes the request. A manager, who is otherwise authorized to
read `secret/foo`, can approve their own request.
```
?

Author: Only if the `self_authorization` property is true (not so in this example).

Reviewer 2: Got it.

```hcl
path "secret/foo" {
    capabilities = ["create", "update"]
    control_group = {
        ttl = "4h"
        factor "tech leads" {
            identity {
                group_names = ["managers", "leads"]
                approvals = 2
            }
        }
        factor "super users" {
            identity {
                group_names = ["superusers"]
                approvals = 1
            }
        }
    }
}
```

The above policy grants `create` and `update` access to `secret/foo` only after
two member of the "managers" or "leads" group and one member of the "superusers"
group authorizes the request. If an authorizer is a member of both the
"managers" and "superusers" group, one authorization for both factors will be

> satisfied.

Reviewer 2: How about this?
```suggestion
The above policy grants `create` and `update` access to `secret/foo` only after several conditions
are met. The required conditions are that:
1. two different users, who can be in either the "managers" or "leads" group (or both), have approved the request;
1. one member of the "superusers" group has authorized the request.

If an authorizer is a member of both the "managers" and "superusers" group, a member of both the the "leads" and "superusers" and "superusers" groups, their authorization counts for one of the two `tech_leads` factor authorizations and also fully satisfies the `super_users` factor.
```

```hcl
path "secret/foo" {
    capabilities = ["create","update","read"]
    control_group = {
        factor "admin" {
            controlled_capabilities = ["update"]
            identity {
                group_names = ["admin"]
                approvals = 1
            }
        }
    }
}
```

The above policy grants `create` and `read` access to `secret/foo` for anyone that has a token
with this policy. It grants `update` access to `secret/foo` only after one member from the

> admin group authorizes the request.

Reviewer 2: This is correct, but actually the permission that people typically lock down further is **read**, not **update** (sometimes both…)

I suggest a more realistic example where only a robot / automation can write the secret, and only a machine identity can read it, except that human ops roles can use a control group authorization to read or update the secret. Only automation, never a human, can create the secret. Plausibly anyone can delete it, because deleting the secret doesn't reveal it and _revealing the secret to a human_ is the thing we want to control.

```hcl
path "kv/*" {
    capabilities = ["create","update","delete","list","sudo"]
    control_group = {
        factor "admin" {
            controlled_capabilities = ["delete","list","sudo"]
            identity {
                group_names = ["admin"]
                approvals = 1
            }
        }
    }
}
path "kv/*" {
    capabilities = ["create"]
    control_group = {
        factor "superuser" {
            identity {
                group_names = ["superuser"]
                approvals = 2
            }
        }
    }
}

```

Because the second path stanza has a control group factor with no `controlled_capabilities` field,
any token with this policy will be required to get 2 approvals from the `superuser` group before executing
any operation against `kv/*`. In addition, by virtue of the `controlled_capabilities` field in the first
path stanza, `delete`,`list`, and `sudo` operations will require an additional approval from the `admin` group.

```hcl
path "kv/*" {

>     capabilities = ["read","list","create"]

Reviewer 2: Again, reading secrets is often a controlled activivity. Keeping things confidential is what OpenBao is all about.

---

_What follows is my personal opinion based on experience working for enterprise clients at ControlPlane and previously in my career._

However, it's OK to include an example where people can read without additional approval. In this concept page we can briefly mention detective / audit controls: maybe anyone can read some less-critical secret, but whenever they do, it triggers rotation within 14 days - or sooner should their employment be terminated.

Just a hint is enough; this page doesn't need to exposit the full set of controls, just indicate that there could be more to it.

Reviewer: @timb-controlplane Not wrong here, but for some context, this document is a revival of one that was present in the `fork-point`, from HashiCorp Vault (Enterprise): https://github.com/openbao/openbao/blob/fork-point/website/content/docs/enterprise/control-groups.mdx -- This example in particular is from there 🤷

I think at some point we might also be happy moving some feedback as issues and deferring them just to keep things moving along in a forward direction, just as a FYI. :-)

Author: yes I agree that fork-point docs could always be improved but are also good enough :-). In this case, we are showing the capability for selective operation control, not the security best practice.

Reviewer 2: Better to skip this or defer to another PR. The perfect is the enemy of the shipped.

- **Unwrap** (`sys/wrapping/unwrap`)**:** Unwrap the token, executing
  the original request and returning a response, once approved. The
  response that is returned will be an original wire-format response;
  it can be used directly with API clients. It is only executed when
  unwrapping is successful.
- **Review** (`sys/control-group/request`)**:** Using the accessor token,
  view information about the original request (entity, path, payload,
  etc).
- **Authorize** (`sys/control-group/authorize`)**:** Using the
  accessor token, approve the request for deferred execution when unwrapped.

## Control-group response-wrapping token creation

Response wrapping is triggered by requesting a path having one or more
applicable control group factors in the policy. At this time, the
original request is stored for deferred execution upon unwrapping.

If a client requests wrapping:

1. The original HTTP response is serialized
2. A new single-use token is generated with the TTL supplied by the factor
3. Internally, the original serialized request is stored.
4. A new response is generated, with the token ID, TTL, and path stored in the
   new response's wrap information object

> 5. The new response is returned to the caller

Reviewer 3: I guess this is a copy&paste oversight from the response wrapping docs.

```suggestion
1. The original HTTP request is serialized
2. A new single-use token is generated with the TTL supplied by the factor
3. Internally, the original serialized request is stored.
4. A response is generated, with the token ID, TTL, and path stored in the
   response's wrap information object
5. The response is returned to the caller
```

Author: adjusted

Note that policies can control minimum/maximum wrapping TTLs; see the [policies
concepts page](policies.mdx) for
more information.

## Control-group factors in ACL policies

Control groups are declared in a control-group stanza of an ACL policy.

> ```

Reviewer 3: ```suggestion
```hcl
```

		self_authorization = false
	  }
	}
  }
}
```

Control Groups can verify the following factors:

- `Identity Groups` - Require an authorizer to be in a specific set of
  identity groups. The `self_authorization` property indicates
  whether the original requester can self-authorize if they belong to
  the control group (default is false).

### Controlled capabilities

Control group factors can be configured to trigger the control group workflow
on specific capabilities. This is done with the `controlled_capabilities` field.
Not specifying the `controlled_capabilities` field will necessitate the factor to be
checked for all operations to the specified policy path. The `controlled_capabilities`
field can differ per factor, so that different factors can be required for different
operations.

Finally, the capabilities in the `controlled_capabilities` stanza must be a subset of the
`capabilities` specified in the policy itself. For example, a policy giving only `read` access to
the path `secret/foo` cannot specify a control group factor with `list` as a controlled capability.

Please see the following section for example ACL Policies.

### Sample ACL policies

> ```

Reviewer 3: ```suggestion
```hcl
```

path "secret/foo" {
    capabilities = ["read"]
    control_group = {
        factor "ops_manager" {
            identity {
                group_names = ["managers"]
                approvals = 1
            }
        }
    }
}
```

The above policy grants `read` access to `secret/foo` only after one member of
the "managers" group authorizes the request.

> ```

Reviewer 3: ```suggestion
```hcl
```

path "secret/foo" {
    capabilities = ["create", "update"]
    control_group = {
        ttl = "4h"
        factor "tech leads" {
            identity {
                group_names = ["managers", "leads"]
                approvals = 2
            }
        }
        factor "super users" {
            identity {
                group_names = ["superusers"]
                approvals = 1
            }
        }
    }
}
```

The above policy grants `create` and `update` access to `secret/foo` only after
two member of the "managers" or "leads" group and one member of the "superusers"
group authorizes the request. If an authorizer is a member of both the
"managers" and "superusers" group, one authorization for both factors will be
satisfied.

> ```

Reviewer 3: ```suggestion
```hcl
```

path "secret/foo" {
    capabilities = ["write","read"]
    control_group = {
        factor "admin" {
            controlled_capabilities = ["write"]
            identity {
                group_names = ["admin"]
                approvals = 1
            }
        }
    }
}
```

> The above policy grants `read` access to `secret/foo` for anyone that has a vault token

Reviewer 3: ```suggestion
The above policy grants `read` access to `secret/foo` for anyone that has a token
```

with this policy. It grants `write` access to `secret/foo` only after one member from the
admin group authorizes the request.

> ```

Reviewer 3: ```suggestion
```hcl
```

path "kv/*" {
    capabilities = ["create", "update","delete","list","sudo"]
    control_group = {
        factor "admin" {
            controlled_capabilities = ["delete","list","sudo"]
            identity {
                group_names = ["admin"]
                approvals = 1
            }
        }
    }
}
path "kv/*" {
    capabilities = ["create"]
    control_group = {
        factor "superuser" {
            identity {
                group_names = ["superuser"]
                approvals = 2
            }
        }
    }
}

```

Because the second path stanza has a control group factor with no `controlled_capabilities` field,
any token with this policy will be required to get 2 approvals from the `superuser` group before executing
any operation against `kv/*`. In addition, by virtue of the `controlled_capabilities` field in the first
path stanza, `delete`,`list`, and `sudo` operations will require an additional approval from the `admin` group.

> ```

Reviewer 3: ```suggestion
```hcl
```

path "kv/*" {
    capabilities = ["read", "list", "create"]
    control_group = {
        controlled_capabilities = ["read"]
        factor "admin" {
            identity {
                group_names = ["admin"]
                approvals = 1
            }
        }
        factor "superuser" {
            controlled_capabilities = ["create"]
            identity {
                group_names = ["superuser"]
                approvals = 1
            }
        }
    }
}
```

In this case, `read` will require one admin approval and `create` will require

> one superuser approval and one admin approval. `List` will require no extra approvals

Reviewer 3: ```suggestion
one superuser approval and one admin approval. `list` will require no extra approvals
```

from any of the control group factors, and a token with this policy will not be required

> to go through the control group workflow in order to execute a read operation against `kv/*`.

Reviewer 3: > In this case, `read` will require one admin approval [...] and a token with this policy will not be required to go through the control group workflow in order to execute a read operation against `kv/*`.

Isn't that self-contradicting or am I misunderstanding something here?

Author: ah, Vault Enterprise docs at fork-point have a `controlled_capabilities` attribute above the factor stanza which is merged with the same attribute within the factor.  I did not implement this! For now, adjusting the example accordingly.

## API

Control Groups can be managed over the HTTP API. Please see

> [Control Groups API](/vault/api-docs/system/control-group) for more details.

Reviewer 3: ```suggestion
[Control Groups API](/api-docs/system/control-group) for more details.
```

Author: this link doesn't seem to compile so I just removed

  the control group (default is false).

### Controlled capabilities

Control group factors can be configured to trigger the control group
workflow on specific capabilities. This is done with the
`controlled_capabilities` field.  Not specifying the
`controlled_capabilities` field will necessitate the factor to be
checked for all operations to the specified policy path. The
`controlled_capabilities` field can differ per factor, so that
different factors can be required for different operations.

Finally, the capabilities in the `controlled_capabilities` stanza must
be a subset of the `capabilities` specified in the policy itself. For
example, a policy giving only `read` access to the path `secret/foo`
cannot specify a control group factor with `list` as a controlled
capability.

Please see the following section for example ACL Policies.

### Sample policies

```hcl
path "secret/foo" {
    capabilities = ["read"]
    control_group = {
        factor "ops_manager" {
            identity {
                group_names = ["managers"]
                approvals = 1
            }
        }
    }
}
```

The above policy grants `read` access to `secret/foo` only after one member of

> the "managers" group authorizes the request.

Reviewer 2: Try this.
```suggestion
The above policy grants `read` access to `secret/foo` only after one member of the "managers" group authorizes the request. If someone is a manager themselves, a different manager must approve it
(because `self_authorization` defaults to false).
```

Author: sounds good, applied it.
